analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs

```diff
@@ -141,6 +143,48 @@ private static NumberConstraint CalculateMultiply(NumberConstraint left, NumberC
         }
     }
 
+    private static BigInteger? CalculateXorMin(NumberConstraint left, NumberConstraint right)
+    {
+        if (left.IsPositive && right.IsPositive)
+        {
+            return left.CanOverlap(right) ? 0 : Min(left.Min - right.Max, right.Min - left.Max);
+        }
+        else if (left.IsNegative && right.IsNegative)
+        {
+            return left.CanOverlap(right) ? 0 : Min(left.Max - right.Min, right.Max - left.Min);
+        }
+        return null;
+
+        static BigInteger? Min(BigInteger? first, BigInteger? second)
+        {
+            if (first is null)
+            {
+                return second;
+            }
+            else if (second is null)
+            {
+                return first;
+            }
+            else
+            {
+                return BigInteger.Min(first.Value, second.Value);
+            }
+        }
+    }
+
+    private static BigInteger? CalculateXorMax(NumberConstraint left, NumberConstraint right)
+    {
+        if ((left.IsPositive && right.IsPositive) || (left.IsPositive && right.CanBePositive) || (right.IsPositive && left.CanBeNegative))
+        {
+            return left.Max.HasValue && right.Max.HasValue ? PositiveMagnitude(left.Max.Value | right.Max.Value) : null;
+        }
+        else if ((left.IsPositive && right.IsNegative) || (left.IsNegative && right.IsPositive))
+        {
+            return -1;
+        }
+        return null;
```

Comment:
This logic confuses me. Shouldn't it be symmetric? From my understanding the following should be correct:
```suggestion
        if ((left.IsPositive && right.IsNegative) || (left.IsNegative && right.IsPositive))
        {
            return -1;
        }
        else if (left.Max.HasValue && right.Max.HasValue)
        {
            return PositiveMagnitude(left.Max.Value | right.Max.Value);
        }
        else
        {
            return null;
        }
```

Reply:
Try this, it's the opposite to my suggested min:
```
        if ((left.IsPositive && right.IsNegative) || (left.IsNegative && right.IsPositive))
        {
            return -1;
        }
        else if ((left.IsPositive || right.IsPositive) && left.Max.HasValue && right.Max.HasValue)
        {
            return PositiveMagnitude(BigInteger.Max(left.Max.Value, right.Max.Value));
        }
        else if ((left.IsNegative || right.IsNegative) && left.Min.HasValue && right.Min.HasValue)
        {
            return PositiveMagnitude(BigInteger.Max(BigInteger.Abs(left.Min.Value), BigInteger.Abs(right.Min.Value)));
        }
        else if (left.Min.HasValue && left.Max.HasValue && right.Min.HasValue && right.Max.HasValue)
        {
            return PositiveMagnitude(BigInteger.Max(
                BigInteger.Max(left.Max.Value, right.Max.Value),
                BigInteger.Max(BigInteger.Abs(left.Min.Value), BigInteger.Abs(right.Min.Value))));
        }
        else
        {
            return null;
        }
        ```

Reply:
Tried that too, and it breaks for the following test case:
```cs
[DataRow("i ==  4 && j >= -6", null, null)]  // exact range: -8, null
Expected SETestContext.CreateCS(code, "int i, int j").Validator.TagValue("Value") 
to have constraints {NotNull}, but additional {Number from * to 7} are present. 
Actual are {NotNull, Number from * to 7}.
```

Counterexample: `i = 4 = 0b00000100 and j = 11 = 0b00001011 => i ^ j = 0b00001111 =15 > 7` 

There is something wrong in the branch that returns 7:
```cs
else if ((left.IsNegative || right.IsNegative) && left.Min.HasValue && right.Min.HasValue)
        {
            return PositiveMagnitude(BigInteger.Max(BigInteger.Abs(left.Min.Value), BigInteger.Abs(right.Min.Value)));
        }
```

So, for max too, I would revert the code to the one surely working, and not invest more time in trying to support these cases, unless there is proven usefulness in doing so.

Also, I am not sure that `return left.Max.HasValue && right.Max.HasValue ? PositiveMagnitude(BigInteger.Max(left.Max.Value, right.Max.Value)) : null;` works correctly for negative numbers, so I am keeping `return left.Max.HasValue && right.Max.HasValue ? PositiveMagnitude(left.Max.Value | right.Max.Value) : null;`, for which I have strong intuition than it does and that works on all UTs.

Reply:
I cannot reproduce the failing test case. Replacing the current implementation with my suggestion improves 5 test cases.